pkg/ebpf/uprobes/attacher.go

```diff
@@ -0,0 +1,721 @@
+// for API compatibility with processMonitor
+func (ua *UprobeAttacher) handleProcessStart(pid uint32) {
+	_ = ua.AttachPIDWithOptions(pid, false) // Do not try to attach to libraries on process start, it hasn't loaded them yet
+}
+
+// handleProcessExit is called when a process finishes, wraps DetachPID but ignoring the error
+// for API compatibility with processMonitor
+func (ua *UprobeAttacher) handleProcessExit(pid uint32) {
+	_ = ua.DetachPID(pid)
+}
+
+func (ua *UprobeAttacher) handleLibraryOpen(event *ebpf.DataEvent) error {
+	defer event.Done()
+
+	libpath := sharedlibraries.ToLibPath(event.Data)
+	path := sharedlibraries.ToBytes(&libpath)
+
+	err := ua.AttachLibrary(string(path), libpath.Pid)
+
+	return err
+}
+
+// AttachLibrary attaches the probes to the given library, opened by a given PID
+func (ua *UprobeAttacher) AttachLibrary(path string, pid uint32) error {
+	if (ua.config.ExcludeTargets&ExcludeSelf) != 0 && int(pid) == os.Getpid() {
+		return ErrSelfExcluded
+	}
+
+	matchingRules := ua.getRulesForLibrary(path)
+	if len(matchingRules) == 0 {
+		return ErrNoMatchingRule
+	}
+
+	registerCB := func(path utils.FilePath) error {
+		return ua.attachToBinary(path, matchingRules)
+	}
+	unregisterCB := func(path utils.FilePath) error {
+		return ua.detachFromBinary(path)
+	}
+
+	return ua.fileRegistry.Register(path, pid, registerCB, unregisterCB)
+}
+
+// getRulesForLibrary returns the rules that match the given library path
+func (ua *UprobeAttacher) getRulesForLibrary(path string) []*AttachRule {
+	var matchedRules []*AttachRule
+
+	for _, rule := range ua.config.Rules {
+		if rule.matchesLibrary(path) {
+			matchedRules = append(matchedRules, rule)
+		}
+	}
+	return matchedRules
+}
+
+// getRulesForExecutable returns the rules that match the given executable
+func (ua *UprobeAttacher) getRulesForExecutable(path string) []*AttachRule {
+	var matchedRules []*AttachRule
+
+	for _, rule := range ua.config.Rules {
+		if rule.matchesExecutable(path) {
+			matchedRules = append(matchedRules, rule)
+		}
+	}
+	return matchedRules
+}
+
+// getExecutablePath resolves the executable of the given PID looking in procfs. Automatically
+// handles delays in procfs updates. Will return an error if the path cannot be resolved
+func (ua *UprobeAttacher) getExecutablePath(pid uint32) (string, error) {
+	pidAsStr := strconv.FormatUint(uint64(pid), 10)
+	exePath := filepath.Join(ua.config.ProcRoot, pidAsStr, "exe")
+
+	var binPath string
+	err := errors.New("iteration start")
+	end := time.Now().Add(procFSUpdateTimeout)
+
+	for err != nil && end.After(time.Now()) {
+		binPath, err = os.Readlink(exePath)
+		if err != nil {
+			time.Sleep(time.Millisecond)
+		}
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	return binPath, nil
+}
+
+// AttachPID attaches the corresponding probes to a given pid
+func (ua *UprobeAttacher) AttachPID(pid uint32) error {
+	return ua.AttachPIDWithOptions(pid, true)
```

Comment:
nit: using `true` and `false` here is not intuitive, and forces the reader to move to the documentation/declaration of `AttachPIDWithOptions`
I'd suggest to make consts as aliases like `const attachToLibs = true` and `const dontAttachToLibs = false`